### ISSUE TITLE
Thread safety in ProxyManager.cpp

### ISSUE THREAD

**Issue by jhughes-mw** [mathworks/libmexclass#11]
It's possible that a proxy could get lost if this this function is called from two threads. They'd both get the same ID and the second would get overridden.
``` 
ID ProxyManager::manageProxy(std::shared_ptr<Proxy> proxy) {
    const ID proxy_id = this->current_proxy_id;
    this->proxy_map[proxy_id] = proxy;
    this->current_proxy_id++;
    return proxy_id;
}
```

**Comment by jhughes-mw** [comment]
Also in the singleton 
```
if (!instance) {
... 
}
```
Could get hit by two threads. Can just use `static const Object` initialization since it's thread safe. Not sure if it's important to have that instance owned by the class or not. Maybe just an implementation detail.

**Comment by jhughes-mw** [comment]
Closed by: bdae1b6437f8e73672996da3b3097301e26dc716
